modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -0,0 +1,321 @@
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        // wait for the cluster state to be recovered
+        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            return;
+        }
+
+        final boolean prevIsMaster = this.isMaster;
+        if (prevIsMaster != event.localNodeMaster()) {
+            this.isMaster = event.localNodeMaster();
+            if (this.isMaster) {
+                // we weren't the master, and now we are
+                onBecomeMaster();
+            } else {
+                // we were the master, and now we aren't
+                cancelJob();
+            }
+        }
+    }
+
+    @Override
+    public void close() {
+        SchedulerEngine engine = scheduler.get();
+        if (engine != null) {
+            engine.stop();
+        }
+    }
+
+    @Override
+    public void triggered(SchedulerEngine.Event event) {
+        if (event.getJobName().equals(DATA_LIFECYCLE_JOB_NAME)) {
+            logger.trace("DLM job triggered: " + event.getJobName() + ", " + event.getScheduledTime() + ", " + event.getTriggeredTime());
+            run(clusterService.state());
+        }
+    }
+
+    /**
+     * Iterates over the DLM managed data streams and executes the needed operations
+     * to satisfy the configured {@link org.elasticsearch.cluster.metadata.DataLifecycle}.
+     */
+    // default visibility for testing purposes
+    void run(ClusterState state) {
+        for (DataStream dataStream : state.metadata().dataStreams().values()) {
+            if (dataStream.getLifecycle() == null) {
+                continue;
+            }
+
+            IndexMetadata writeIndex = state.metadata().index(dataStream.getWriteIndex());
+            if (writeIndex != null && isManagedByDLM(dataStream, writeIndex)) {
+                RolloverRequest rolloverRequest = defaultRolloverRequestSupplier.apply(dataStream.getName());
+                transportActionsDeduplicator.executeOnce(
+                    rolloverRequest,
+                    ActionListener.noop(),
+                    (req, reqListener) -> rolloverDataStream(rolloverRequest, reqListener)
+                );
+            }
+
+            TimeValue retention = getRetentionConfiguration(dataStream);
+            if (retention != null) {
+                List<Index> backingIndices = dataStream.getIndices();
+                // we'll look at the current write index in the next run if it's rolled over (and not the write index anymore)
+                for (int i = 0; i < backingIndices.size() - 1; i++) {
+                    IndexMetadata backingIndex = state.metadata().index(backingIndices.get(i));
+                    if (backingIndex == null || isManagedByDLM(dataStream, backingIndex) == false) {
+                        continue;
+                    }
+
+                    TimeValue indexLifecycleDate = getCreationOrRolloverDate(dataStream.getName(), backingIndex);
```

Comment:
Hm, I was wondering if it would be better to call this `rolloverDate`, at this point in the code we can only encounter rolled over indices, right?

Reply:
I'd prefer to not make that assumption as the `modify datastream` API could be used to bring any index into the data stream (e.g. one that was never rolled over - at which point we'd take that index's creation data into consideration)